### DISCUSSION TITLE
Simple UIKit Navigation

### DISCUSSION THREAD

**Discussion by michaelfransen** [pointfreeco/swift-composable-architecture#1497] [Q&A] [answered]
I've recently started exploring TCA and am getting hung up on simple UIKit navigation. I started by building a simple feature contained in a UIViewController, and I had no problems with that. Now I am trying to create a main view controller that will link to my existing feature view controller and will have other features that will link out from my main view controller via just having a bunch of UIButtons on the screen for each feature. I am struggling with the setup for this main view controller. I have a reducer and state for the main view controller. The state contains a property for my feature state. Where I am confused is when it comes to actions. I want it so when one UIButton is tapped, that feature is launched modally. What should the actions look like for this main view controller? It is my understanding that in the actions enum, you should have your child feature actions as a case, but what is used to trigger the actual display? Should one make another action called `case presentEchoFeature`? The tick tac toe example app just has state as an enum and then they just swap the contained view controller. 

```
struct TCAExample: ReducerProtocol {
    struct State: Equatable {
        var echo: Echo.State?
    }

    enum Action: Equatable {
        case echo(Echo.Action)
    }

    var body: some ReducerProtocol<State, Action> {
        Reduce { state, action in
            switch action {
            case .echo:
                return .none
            }
        }
    }
}
```

**Comment by stephencelis** [comment, marked as answer]
Take a look at the [Tic-Tac-Toe demo](https://github.com/pointfreeco/swift-composable-architecture/tree/86438fad149bcf38103dd1dc2ec9f3fd978c0dfa/Examples/TicTacToe) that comes with the repo. It has both SwiftUI and UIKit implementations that share the same domain code.

The `LoginViewController` has an example of pushing a view controller based off optional state here:

https://github.com/pointfreeco/swift-composable-architecture/blob/86438fad149bcf38103dd1dc2ec9f3fd978c0dfa/Examples/TicTacToe/tic-tac-toe/Sources/LoginUIKit/LoginViewController.swift#L159-L173

The action that kicks off the event happens to be the result of a side effect in the reducer. But if you want to immediately present based off a button tap, you could use that action to set your associated "echo" state to non-`nil`.

**Reply by andriyslyusar** [reply]
Stephen are you planning to address UIKit navigation in the upcoming navigation video series? The approach from Tic-Tac-Toe demo is working, but it does not quite fill as the right approach.

**Comment by michaelfransen** [comment]
Thanks a ton for the response. So this works, but this feels wrong. In my actions I have an action that changes the state, and then I have another action for the child actions. Is this correct? 

``` swift
struct TCAExample: ReducerProtocol {
    struct State: Equatable {
        var echo: Echo.State?
    }

    enum Action: Equatable {
        case dislayEchoFeature
        case echo(Echo.Action)
    }

    var body: some ReducerProtocol<State, Action> {
        Reduce { state, action in
            switch action {
            case .dislayEchoFeature:
                state.echo = .init()
                return .none
            case .echo:
                return .none
            }
        }
    }
}
```

**Reply by tgrapperon** [reply]
Indeed, this usually the way to go with the library. If having all the actions bundled into one enum doesn't fit your coding style, you may want to experiment with sub-enums to structure your domains more tightly. See for example the recent discussion: #1440. I wouldn't advise to do so if you're starting with the library, as it will create a little friction that will probably make things more difficult to grasp. But you can totally switch to this style once you get a better view of the way TCA works.

Please also note that you don't have all your app's actions into one enum[^1], as your features compose other features that are themselves composed, etc. So you usually only have a few top-level actions in a given domain/feature.

[^1]: Well, technically, you have, but it's hidden by composition.

**Reply by michaelfransen** [reply]
Thank you for the response. Greatly appreciated!
